man/sio.3
.\"(c) Copyright 1992, 1993 by Panagiotis Tsirigotis
.\"All rights reserved.  The file named COPYRIGHT specifies the terms
.\"and conditions for redistribution.
.\"
.\" $Id$
.TH SIO 3X "29 May 1992"
.SH NAME
Sread, Sgetc, Srdline, Sfetch, Swrite, Sputc, Sprint, Sprintv, Sdone, Sundo, Stie, Suntie, Sflush, Sclose, Sbuftype, Smorefds, Sgetchar, Sputchar, SIOLINELEN - fast stream I/O
.SH SYNOPSIS
.LP
.nf
.ft B
#include "sio.h"
#include <stdarg.h>
.LP
.ft B
int Sread( fd, buf, nbytes )
int fd ;
char *buf ;
int nbytes ;
.LP
.ft B
int Sgetc( fd )
int fd ;
.LP
.ft B
char *Srdline( fd )
int fd ;
.LP
.ft B
char *Sfetch( fd, length )
int fd ;
long *length ;
.LP
.ft B
int Swrite( fd, buf, nbytes )
int fd ;
char *buf ;
int nbytes ;
.LP
.ft B
int Sputc( fd, c )
int fd ;
char c ;
.LP
.ft B
int Sprint( fd, format [ , ... ] )
int fd ;
char *format ;
.LP
.ft B
int Sprintv( fd, format, ap )
int fd ;
char *format ;
va_list ap ;
.LP
.ft B
int Sdone( fd )
int fd ;
.LP
.ft B
int Sundo( fd, type )
int fd ;
int type ;
.LP
.ft B
int Stie( ifd, ofd )
int ifd, ofd ;
.LP
.ft B
int Suntie( fd )
int fd ;
.LP
.ft B
int Sbuftype( fd, type )
int fd, type ;
.LP
.ft B
int Smorefds()
.LP
.ft B
int Sflush( fd )
int fd ;
.LP
.ft B
int Sclose( fd )
int fd ;
.LP
.ft B
int Sgetchar( fd )
int fd ;
.LP
.ft B
int Sputchar( fd, c )
int fd;
char c ;
.LP
.ft B
int SIOLINELEN( fd )
int fd ;
.SH DESCRIPTION
The \fISIO\fR library provides support
for \fIstream\fR I/O on file descriptors.
The first argument of every function
or macro is a file descriptor. The file descriptor may be used either for
input or for output but not both. Attempting to use a descriptor for
both input and output will cause the call to fail. When you are
done with using a file descriptor, you should inform \fISIO\fR
by invoking \fBSdone()\fR (unless the program is about to
call \fIexit(3)\fR).
You can also use \fBSdone()\fR if
you want to perform a different type of operation on the same
file descriptor (e.g. first you were reading data from the file
descriptor and then you want to write some data).
Another possibility is to do stream I/O at different file offsets
by using \fBSdone()\fR before doing the \fBlseek(2)\fR to the
new file offset.
.LP
I/O operations on different file descriptors do not interfere
(unless the file descriptors refer to the same file, in which case
the results are undefined).
.LP
For disk files I/O always starts at the current file offset.
If that offset is not a multiple of the preferred block size for file
system I/O (the \fIst_blksize\fR field in \fIstruct stat\fR),
performance will not be optimal.
For optimal performance, it is recommended that no I/O operations
(like \fIread(2)\fR or \fIwrite(2)\fR)
are applied to the file descriptor if it is to be used by \fISIO\fR.
.LP
Read I/O is either buffered or is done using memory mapping whenever
that is possible and appropriate.
.LP
The library functions that do stream I/O resemble system calls
(for example \fBSread()\fR resembles \fIread(2)\fR) so that modifying
a program that uses the system calls to use the \fISIO\fR functions
is easy (e.g. just replace \fIread(2)\fR with \fBSread()\fR; the function
signatures as well as the return values are exactly the same).
.LP
.B Sread()
reads \fInbytes\fR bytes from the stream associated with file
descriptor \fIfd\fR into the buffer pointed to by \fIbuf\fR.
.LP
.B Sgetc()
reads a character from the stream
associated with file descriptor \fIfd\fR.
It returns \fBSIO_EOF\fR if the end of file has been reached.
.LP
.B Sgetchar()
(a macro) performs exactly the same function as \fBSgetc()\fR but
it is much faster.
.LP
.B Srdline()
reads a line from the stream
associated with file descriptor \fIfd\fR.
The newline at the end of the line is replaced by a 0 byte. Lines
longer than the maximum line length supported by \fISIO\fR will
have characters deleted.
.LP
.B SIOLINELEN()
(a macro) returns the length of
the line returned by the last call to \fBSrdline()\fR
(the value returned by \fBSIOLINELEN()\fR is valid only after
\fBSrdline()\fR and as long as no other
\fISIO\fR calls are performed on that file descriptor).
.LP
.B Sfetch()
returns a pointer to data coming from the stream
associated with file
descriptor \fIfd\fR. The amount of data available is indicated
by the \fIlength\fR argument. One possible use for this function
is copying of files.
.LP
.B Swrite()
writes \fInbytes\fR bytes to the stream associated with file
descriptor \fIfd\fR from the buffer pointed to by \fIbuf\fR.
.LP
.B Sputc()
writes a single character to the stream
associated with file descriptor \fIfd\fR.
.LP
.B Sputchar()
(a macro) performs exactly the same function as \fBSputc()\fR
but it is much faster.
.LP
.B Sprint()
imitates the behavior of printf(3) as defined in the
ANSI C Standard. There are some limitations. Check the \fBSprint()\fR
man page for more information.
.LP
.B Sprintv()
is the same as \fBSprint()\fR except that it takes a
\fIstdarg\fR argument list.
.LP
.B Sundo()
returns the characters returned by the last call to
\fBSrdline()\fR, \fBSgetc()\fR or \fBSgetchar()\fR to the stream
so that they can be reread. The \fItype\fR argument to \fBSundo()\fR
can be \fBSIO_UNDO_LINE\fR or \fBSIO_UNDO_CHAR\fR depending
on whether the call whose effect needs to be undone was
\fBSrdline()\fR or \fBSgetc()\fR/\fBSgetchar()\fR respectively.
There is no check on
whether the last function invoked on \fIfd\fR was one of the above
and the results are undefined if there is no correspondence
between the \fItype\fR and the last operation on \fIfd\fR.
(i.e. the result is undefined if you try \fBSIO_UNDO_CHAR\fR
and the last operation was not \fBSgetchar()\fR or \fBSgetc()\fR).
.LP
.B Stie()
ties the file descriptor \fIifd\fR to the file descriptor \fIofd\fR.
This means that whenever a \fIread(2)\fR is done on \fIifd\fR, it is
preceded by a \fIwrite(2)\fR on \fIofd\fR.
For filters it is useful to do \fIStie( 0, 1 )\fR to maximize concurrency.
It is also useful to do the same thing when you issue prompts to the
user and you want the user reply to appear on the same line with the
prompt.
\fIifd\fR, \fIofd\fR  will be initialized for input, output respectively
(if any of them is initialized, it must be for the appropriate
stream type (input or output)).
If \fIifd\fR was tied to another file descriptor, the old tie is broken.
.LP
.B Suntie()
undoes the effect of \fBStie()\fR for the specified input file descriptor.
.LP
.B Sbuftype()
determines the buffering type for the output stream associated with
file descriptor \fIfd\fR.
By default output directed to terminals is line buffered, output
directed to file descriptor 2 (standard error) is unbuffered and
everything else is fully buffered.
Possible values for the \fItype\fR argument are
.RS
.TP 15
.B SIO_FULLBUF
for full buffering
.TP
.B SIO_LINEBUF
for line buffering
.TP
.B SIO_NOBUF
for no buffering
.RE
.LP
.B Smorefds()
should be used to inform \fBSIO\fR that the number of available file
descriptors has been increased. \fBSIO\fR uses an array of internal
stream descriptors which are indexed by the file descriptor number. Some
operating systems (ex. SunOS 4.1[.x]) allow the number of available
file descriptors to vary. If that number is increased beyond its initial
value \fBSIO\fR needs to know in order to allocate more stream descriptors.
.LP
.B Sdone()
flushes any buffered output for \fIfd\fR
and releases the \fISIO\fR resources used. \fBSdone()\fR
is useful in case the program needs to reprocess the
data of a file descriptor (assuming the file descriptor corresponds
to a file).  The program can call \fBSdone()\fR,
\fIlseek(2)\fR to the beginning of the file
and then proceed to reread the file.
.LP
.B Sflush()
causes any buffered stream output to be written to the
file descriptor. If its argument is the special value \fBSIO_FLUSH_ALL\fR
then all output streams will be flushed.
.LP
.B Sclose()
closes a file descriptor used for stream I/O, flushes
any buffered output and releases the \fISIO\fR resources used.
.SH EXAMPLES
.LP
The following code implements a (poor) substitute for the tee command
(it copies standard input to a file as well as to standard output).
.ne 10
.RS
.nf
.ft B
#include "sio.h"
.sp .5
main( argc, argv )
	int argc ;
	char *argv[] ;
{
	char *file = (argc > 1) ? argv[ 1 ] : "tee.file" ;
	int fd = creat( file, 0644 ) ;
	long length ;
	char *s ;
.sp .5
	while ( s = Sfetch( 0, &length ) )
	{
		Swrite( 1, s, length ) ;
		Swrite( fd, s, length ) ;
	}
	exit( 0 ) ;
}
.fi
.ft R
.RE
.SH RETURN VALUES
.LP
.B Sread()
returns the number of bytes read on success
(0 means end-of-file)
or \fBSIO_ERR\fR on failure (\fIerrno\fR is set to indicate the error).
.LP
.B Sgetc()
returns the character read on success,
SIO_EOF when the end-of-file is reached,
or \fBSIO_ERR\fR on failure (\fIerrno\fR is set to indicate the error).
.LP
.B Srdline()
returns a pointer to the next line on success.
On failure or when the end-of-file is reached it returns
.SM NULL.
If the end-of-file is reached \fIerrno\fR is set to 0, otherwise it indicates
the error.
.LP
.B Sfetch()
returns a pointer to file data on success.
(the \fIlength\fR argument indicates how many bytes
are available).
On failure or when the end-of-file is reached it returns
.SM NULL.
If the end-of-file is reached \fIerrno\fR is set to 0, otherwise it indicates
the error.
.LP
.B Swrite()
returns the number of bytes written on success
or \fBSIO_ERR\fR on failure (\fIerrno\fR is set to indicate the error).
.LP
.B Sputc()
returns the character it was given as an argument on success
.B Sprint()
returns the number of characters printed on success
or \fBSIO_ERR\fR on failure (\fIerrno\fR is set to indicate the error).
.LP
.B Sdone()
returns \fB0\fR on success
or \fBSIO_ERR\fR on failure (\fIerrno\fR is set to indicate the error).
.LP
.B Sundo()
returns \fB0\fR on success
or \fBSIO_ERR\fR on failure (\fIerrno\fR is set to indicate the error).
.LP
.B Stie()
returns \fB0\fR on success
or \fBSIO_ERR\fR on failure (\fIerrno\fR is set to indicate the error).
.LP
.B Suntie()
returns \fB0\fR on success
or \fBSIO_ERR\fR on failure
(\fIerrno\fR is set to \fBEBADF\fR if there
was no tied file descriptor).
.LP
.B Sbuftype()
returns \fB0\fR on success
or \fBSIO_ERR\fR on failure
(\fIerrno\fR is set to \fBEBADF\fR if this is not an output stream
or to \fBEINVAL\fR if an unknown \fItype\fR is specified).
.LP
.B Smorefds()
returns \fB0\fR on success
or \fBSIO_ERR\fR on failure (because of lack of memory).
.LP
.B Sflush()
returns \fB0\fR on success
or \fBSIO_ERR\fR on failure (\fIerrno\fR is set to indicate the error).
.LP
.B Sclose()
returns \fB0\fR on success
or \fBSIO_ERR\fR on failure (\fIerrno\fR is set to indicate the error).
.LP
.B Sgetchar()
returns the character read on success,
SIO_EOF when the end-of-file is reached,
or \fBSIO_ERR\fR on failure (\fIerrno\fR is set to indicate the error).
.LP
.B Sputchar()
returns the character it was given as an argument on success
or \fBSIO_ERR\fR on failure (\fIerrno\fR is set to indicate the error).
.LP
.B SIOLINELEN()
returns the length of the last line read by \fBSrdline()\fR.
.LP
Attempting a read operation on a descriptor opened for writing or vice
versa will cause the operation to fail with \fIerrno\fR set to \fBEBADF\fR.
.LP
The first \fISIO\fR operation on a descriptor must be a read or write
operation. It cannot be a control operation (like \fBSflush()\fR). Such
an operation will fail with \fIerrno\fR set to \fBEBADF\fR.
.LP
.IP "\fBNOTE 1:\fR" 15
\fBStie()\fR is an input/output operation for the
respective file descriptors, not a control operation. \fBSuntie()\fR
is a control operation.
.IP "\fBNOTE 2:\fR"
\fBSIO_ERR\fR is defined to be \fB-1\fR.
.SH "SEE ALSO"
.LP
Sprint(3)
.SH BUGS
.LP
If the operating system does not provide for invocation of a
finalization function upon exit, the program will have to
explicitly flush all output streams.
The following operating systems provide such a facility:
SunOS 4.x, Ultrix 4.x.
.LP
Socket file descriptors can be used for input as well as output but
\fBSIO\fR does not support this.
.LP
The current implementation will not try to use memory mapping to
read a file if the file offset is not 0 (it will use buffered I/O instead).
.LP
Pointers returned by \fBSfetch()\fR point to read-only memory.
Attempting to modify this memory will result in a segmentation
violation.


CHANGELOG

This file describes the changes to xinetd. The base version is 2.0.0.

2.0.0: (not released)

2.0.1: (not released)

2.0.2:
	Changes to the Makefile.
	Trivial mods to the usage() function

2.0.3:
	Changes to the Makefile.
	Fixed a bug in the usage() function

2.0.4:
	Distribution versions of the Makefile no longer contain any 
	references to options.opt

2.0.5:
	1) Clarified what an "unlisted RPC service" is in the man page
	2) Fixed a bug in remote_address_check which caused access to be denied
		for all hosts if no_access was set and only_from was not set.
	3) Fixed a bug which caused arbitrary syslog levels if the log_type
		was specified as SYSLOG in the "defaults" entry

2.0.6:
	bug fix in child.c: replaced strx_sprint with strx_print when naming
	interceptor processes


-------------------------------------------------------------------------------

2.1.0: (not released)
	This was mostly a clean-up of 2.0
	List of changes:

	A. The man page has been split into 3 parts:
			xinetd.man			:	man page describing the program
			xinetd.conf.man	:	man page describing the configuration file 
										This file now includes figures about the
										overhead of interception.
			xinetd.log.man		:	man page describing the log file
	B. New service attributes
			rpc_number			: to support unlisted RPC services
			nice					: to set the nice value of forked servers
	C. The IDONLY service flag was added
	D. Now uses the timer library unless NO_TIMERS is defined. If NO_TIMERS
		is not defined, the following are also available:
			a) a new option, -cc, to do periodic consistency checks
			b) timeout for reconfiguration
	E.	Configuration file man page now mentions that access control is 
		based on IP-address instead of domain address.
	F.	The interception code now sets the TCP_NODELAY option.
	G. The timeout when contacting a remote identification server is
		configurable both when the service request has been accepted and 
		when the request is rejected. In the former case it defaults to
		infinity while in the latter it defaults to 30 sec.
	H. The log line ids are now constants in an include file.
	I. The fsma library is no longer used.
	J. Most structure fields were renamed; certain functions were renamed too.
		Access to structure fields is now via macros.
	K. A bug in not restoring the number of descriptors to the soft limit
		was fixed.
	L. The -pid option works
	M. New internal services:
			"servers"		: lists active servers
			"services"		: lists active services

2.1.1:
	xinetd produces more meaningful messages if it can't start logging

2.1.2:
	Fixed 2 bugs:
		1) xinetd would request a port number for unlisted RPC services
			One could get around this by simply specifying the port attribute
			for the service; the specified value would be ignored anyway.
		2) xinetd initialization might fail sometimes in function msg_init()
			syslog logging was specified (i.e. the option -syslog was used)
			The cause of this bug was that the 
				xlog_control( xlog, XLOG_GETFD, &fd ) 
			operation will not fail for xlogs connected to syslog and
			the value of 'fd' would be used in the subsequent fcntl(). That
			value is arbitrary since 'fd' is a local variable.
			If 'fd' did not happen to refer to an open descriptor, the program 
			would terminate since the fcntl() would fail (btw, the success
			of the fcntl() call would be harmless if 'fd' happened to
			refer to an open descriptor).

2.1.3:
	Bug fix:
		only_from/no_access addresses would get inverted on little-endian
		machines when such addresses were specified using the numeric notation
		(for example, 128.138.45.3). This bug was in the numeric_addr function
		which did not convert the result of inet_addr to host-byte-order.
		A work around for the bug would be to use the factorized address 
		notation (for example, 128.138.45.{3})

2.1.4:
	Bug fixes:
		1) in dgram_echo(), sin_len was not being set before the invocation of
			recvfrom
		2) in finger_shutdown(), it was possible for Srdline() to return NULL
			(if the remote end would close the socket without sending anything).
			If the RECORD option was set in the log_on_failure flags, this
			would cause the forked xinetd process which did the recording to
			die since it would try to dereference a NULL pointer.


2.1.5:
	Bug fixes:
		1) in exec_server() service descriptor might be closed when
		   execing server. This made it impossible to start servers
			for 'nowait' services.  The bug occurred only Ultrix version
			4.3a or (probably) later (bug discovered and fix provided
			by doug@seas.smu.edu)
		
		2) for systems that supported supplementary group id's, the
			set_credentials() function did not set those group id's
			(they were being inherited from xinetd). Now initgroups(3)
			is called to set the supplementary group id's properly.
			(bug discovered and fix provided by maf+@osu.edu)

2.1.6:
	Bug fixes:
		1) xinetd will crash after reconfiguration if there is a running
			server for a service that was removed from the configuration and
			which logs on exit.
		2) xinetd forked process falls in infinite loop if identd server
			sends a reply that is missing the ending CR-LF.
			(bug discovered and fix provided by Laurent.Wacrenier@gin.obspm.fr).

	We also change the LOGUSER_SUCCESS_TIMEOUT constant in config.h from
	0 seconds (i.e. infinite timeout) to 30 seconds. This avoids infinite
	waits in case the remote host does not send a RST reply when we attempt
	to connect to the IDENTD port, and there is no server listening at
	that port.


2.1.7:
	Bug fix: the HOST flag in the 'log_on_success' attribute was ignored;
				the code was incorrectly checking if the HOST flag was
				set in the log_on_failure attribute
				(bug discovered by frolich@corrine.cpc.cs.ucf.edu)

2.1.8.1: 
	Added support for TCP redirect to a remote host.
		--bbraun

2.1.8.2: 
	Added support for binding to specific interface.
		--bbraun

2.1.8.3:
	Changed redirect so that a remote port can be specified as well.

xinetd/sio/sio.c
/*
 * (c) Copyright 1992, 1993 by Panagiotis Tsirigotis
 * (c) Portions Copyright 1998-2001 by Rob Braun
 * All rights reserved.  The file named COPYRIGHT specifies the terms 
 * and conditions for redistribution.
 */

#include "config.h"
#ifndef linux
#include <sys/types.h>
#endif

#ifdef _APPLE_
#undef HAVE_MMAP
#endif

#include <limits.h>
#include <stdlib.h>
#include <unistd.h>

#include "sio.h"
#include "impl.h"

/*
 * SIO WRITE FUNCTIONS: Swrite, Sputc
 */

/*
 * Stream write call: arguments same as those of write(2)
 */
ssize_t Swrite( int fd, const char *addr, size_t nbytes )
{
	__sio_descriptor_t *dp ;
	__sio_od_t *odp ;
	size_t b_transferred ;
	size_t b_avail ;
	ssize_t total_b_transferred ;
	ssize_t b_written ;
	ssize_t b_in_buffer ;

	if ( nbytes > SSIZE_MAX )
		return( SIO_ERR );
	if( sio_setup( fd, &dp, __SIO_OUTPUT_STREAM ) == SIO_ERR )
		return( SIO_ERR );
	odp = ODP( dp ) ;
	ASSERT( odp->start <= odp->nextb && odp->nextb <= odp->buf_end ) ;

	b_avail = odp->buf_end - odp->nextb ;
	b_transferred = MIN( nbytes, b_avail ) ;
	sio_memcopy( addr, odp->nextb, b_transferred ) ;
	odp->nextb += b_transferred ;

	/*
	 * check if we are done
	 */
	if ( b_transferred == nbytes )
		return( b_transferred ) ;

	/*
	 * at this point we know that the buffer is full
	 */
	b_in_buffer = odp->buf_end - odp->start ;
	b_written = __sio_writef( odp, fd ) ;
	if ( b_written != b_in_buffer )
		return( (b_written >= (ssize_t)nbytes) ? (ssize_t)nbytes : b_written ) ;
	
	total_b_transferred = b_transferred ;
	addr += b_transferred ;
	nbytes -= b_transferred ;

	for ( ;; )
	{
		b_transferred = MIN( nbytes, odp->buffer_size ) ;
		sio_memcopy( addr, odp->nextb, b_transferred ) ;
		odp->nextb += b_transferred ;
		nbytes -= b_transferred ;
		if ( nbytes == 0 )
		{
			total_b_transferred += b_transferred ;
			break ;
		}
		/*
		 * the buffer is full
		 */
		b_written = __sio_writef( odp, fd ) ;
		if ( b_written != (ssize_t)odp->buffer_size )
		{
			if ( b_written != SIO_ERR )
			{
				total_b_transferred += b_written ;
				odp->nextb += b_written ;
			}
			break ;
		}
		/*
		 * everything is ok
		 */
		total_b_transferred += b_transferred ;
		addr += b_transferred ;
	}
	return( total_b_transferred ) ;
}


/*
 * Add a character to a file
 */
static int Sputc( int fd, char c )
{
	__sio_descriptor_t *dp ;
	__sio_od_t *odp ;

	if( sio_setup( fd, &dp, __SIO_OUTPUT_STREAM) == SIO_ERR )
		return( SIO_ERR );
	odp = ODP( dp ) ;
	ASSERT( odp->start <= odp->nextb && odp->nextb <= odp->buf_end ) ;

	/*
	 * The following is a weak check since we should really be
	 * checking that nextb == buf_end (it would be an error for
	 * nextb to exceed buf_end; btw, the assertion above, when
	 * enabled makes sure this does not occur).
	 *
	 * NOTE: __sio_writef NEVER uses data beyond the end of buffer.
	 */
	if ( odp->nextb >= odp->buf_end )
	{
		int b_in_buffer = odp->buf_end - odp->start ;

		/*
		 * There is nothing we can do if __sio_writef does not manage
		 * to write the whole buffer
		 */
		if ( __sio_writef( odp, fd ) != b_in_buffer )
			return( SIO_ERR ) ;
	}
	*odp->nextb++ = c ;
	if ( __SIO_MUST_FLUSH( *odp, c ) && __sio_writef( odp, fd ) == SIO_ERR )
		return( SIO_ERR ) ;
	return ( c ) ;
}



/*
 * SIO READ FUNCTIONS
 */


/*
 * Read a line from a file
 * Returns a pointer to the beginning of the line or NULL
 */
char *Srdline( int fd )
{
	__sio_descriptor_t *dp ;
	__sio_id_t *idp ;
	char *cp ;
	char *line_start ;
	size_t b_left ;
	ssize_t extension ;

	if( sio_setup( fd, &dp, __SIO_INPUT_STREAM ) == SIO_ERR )
		return( NULL );
	idp = IDP( dp ) ;
	ASSERT( idp->start <= idp->nextb && idp->nextb <= idp->end ) ;

#ifdef HAVE_MMAP
	if ( idp->memory_mapped && __sio_switch( idp, fd ) == FAILURE )
		return( NULL ) ;
#endif

	b_left = idp->end - idp->nextb ;
	/*
	 * Look for a '\n'. If the search fails, extend the buffer
	 * and search again (the extension is performed by copying the
	 * bytes that were searched to the auxiliary buffer and reading 
	 * new input in the main buffer).
	 * If the new input still does not contain a '\n' and there is
	 * more space in the main buffer (this can happen with network
	 * connections), read more input until either the buffer is full
	 * or a '\n' is found.
	 * Finally, set cp to point to the '\n', and line_start to
	 * the beginning of the line
	 */
	if ( b_left && ( cp = sio_memscan( idp->nextb, b_left, '\n' ) ) != NULL )
	{
		line_start = idp->nextb ;
		idp->nextb = cp + 1 ;
	}
	else
	{
		extension = __sio_extend_buffer( idp, fd, b_left ) ;
		if ( extension > 0 )
		{
			ASSERT( idp->start <= idp->nextb && idp->nextb <= idp->end ) ;

			line_start = idp->start ;
			cp = sio_memscan( idp->nextb, (size_t)extension, '\n' ) ;
			if ( cp != NULL )
				idp->nextb = cp + 1 ;
			else
				for ( ;; )
				{
					idp->nextb = idp->end ;
					extension = __sio_more( idp, fd ) ;
					if ( extension > 0 )
					{
						cp = sio_memscan( idp->nextb, (size_t)extension, '\n' ) ;
						if ( cp == NULL )
							continue ;
						idp->nextb = cp + 1 ;
						break ;
					}
					else
					{
						/*
						 * If there is spare room in the buffer avoid trashing
						 * the last character
						 */
						if ( idp->end < &idp->buf[ idp->buffer_size ] )
							cp = idp->end ;
						else
							cp = &idp->buf[ idp->buffer_size - 1 ] ;
						break ;
					}
				}
		}
		else					/* buffer could not be extended */
			if ( b_left == 0 )
			{
				/*
				 * Set errno to 0 if EOF has been reached
				 */
				if ( extension == 0 )
					errno = 0 ;
				return( NULL ) ;
			}
			else
			{
				line_start = idp->start ;
				cp = idp->end ;
				/*
				 * By setting idp->nextb to be equal to idp->end,
				 * subsequent calls to Srdline will return NULL because
				 * __sio_extend_buffer will be invoked and it will return 0.
				 */
				idp->nextb = idp->end ;
			}
	}
	*cp = NUL ;
	idp->line_length = cp - line_start ;
	return( line_start ) ;
}



/*
 * SIO CONTROL FUNCTIONS
 */


/*
 * Flush the buffer associated with the given file descriptor
 * The special value, SIO_FLUSH_ALL flushes all buffers
 *
 * Return value:
 *			0 :  if fd is SIO_FLUSH_ALL or if the flush is successful
 *		SIO_ERR: if fd is not SIO_FLUSH_ALL and
 *								the flush is unsuccessful
 *							or the descriptor is not initialized or it is not 
 *								an output descriptor
 */
int Sflush( int fd )
{
   __sio_descriptor_t *dp ;
   int b_in_buffer ;

   if ( fd == SIO_FLUSH_ALL )
   {
      for ( fd = 0, dp = __sio_descriptors ;
				fd < __sio_n_descriptors ;
				dp++, fd++ )
         if ( DESCRIPTOR_INITIALIZED( dp ) &&
							dp->stream_type == __SIO_OUTPUT_STREAM )
            (void) __sio_writef( ODP( dp ), fd ) ;
      return( 0 ) ;
   }
   else
   {
      if ( fd >= __sio_n_descriptors ) {
         errno = EBADF ;
         return( SIO_ERR ) ;
      }

      dp = &__sio_descriptors[ fd ] ;

      if( (! DESCRIPTOR_INITIALIZED( dp )) || 
            (dp->stream_type != __SIO_OUTPUT_STREAM) )
         return( SIO_ERR );

      b_in_buffer = ODP( dp )->nextb - ODP( dp )->start ;
      if ( __sio_writef( ODP( dp ), fd ) != b_in_buffer )
         return( SIO_ERR ) ;
      else
         return( 0 ) ;
   }
}


/*
 * Close the file descriptor. This call is provided because
 * a file descriptor may be closed and then reopened. There is
 * no easy way for SIO to identify such a situation, so Sclose
 * must be used.
 *
 * Sclose invokes Sdone which finalizes the buffer.
 * There is no SIO_CLOSE_ALL value for fd because such a thing
 * would imply that the program will exit very soon, therefore
 * the closing of all file descriptors will be done in the kernel
 * (and the finalization will be done by the finalization function
 * NOTE: not true if the OS does not support a finalization function)
 *
 * There is no need to invoke SETUP; Sdone will do it.
 */
int Sclose( int fd )
{
	if ( __SIO_FD_INITIALIZED( fd ) ) {
		if ( Sdone( fd ) == SIO_ERR ) {
			close( fd ); 
			return( SIO_ERR ) ;
		}
	}
	return( close( fd ) ) ;
}


/*
 * Changes the type of buffering on the specified descriptor.
 * As a side-effect, it initializes the descriptor as an output stream.
 */
int Sbuftype( int fd, int type )
{
	__sio_descriptor_t *dp ;

	/*
	 * Check for a valid type
	 */
	if ( type != SIO_LINEBUF && type != SIO_FULLBUF && type != SIO_NOBUF )
	{
		errno = EINVAL ;
		return( SIO_ERR ) ;
	}

	if( sio_setup( fd, &dp, __SIO_OUTPUT_STREAM ) == SIO_ERR )
		return( SIO_ERR );
	ODP( dp )->buftype = type ;
	return( 0 ) ;
}


#ifndef sio_memscan

static char *sio_memscan( const char *from, size_t how_many, char ch )
{
   char *p ;
   char *last = from + how_many ;

   for ( p = from ; p < last ; p++ )
      if ( *p == ch )
         return( p ) ;
      return( 0 ) ;
}

#endif	/* sio_memscan */


#ifdef NEED_MEMCOPY

void __sio_memcopy( const char *from, char *to, size_t nbytes )
{
   while ( nbytes-- )
      *to++ = *from++ ;
}

#endif /* NEED_MEMCOPY */

int sio_setup(int fd, __sio_descriptor_t **dp, unsigned int type)
{
   if ( fd >= __sio_n_descriptors ) {
      if( Smorefds(fd) != 0 ) {
         errno = EBADF;
         return(SIO_ERR);
      }
   }

   *dp = &__sio_descriptors[ fd ];

   if( DESCRIPTOR_INITIALIZED( *dp ) ) {
      if( (*dp)->stream_type != type ) {
         errno = EBADF;
         return( SIO_ERR );
      }
   } else if( __sio_init(*dp, fd, type) == SIO_ERR )
      return( SIO_ERR );

   return( 0 );
}

/*
 * The Sputchar function depends on the fact that the fields
 *                              nextb, buf_end, end
 * are 0 if a stream descriptor is not being used or has not yet been
 * initialized.
 * This is true initially because of the static allocation of the
 * descriptor array, and Sdone must make sure that it is true
 * after I/O on a descriptor is over.
 */
int Sputchar( int fd, char c )
{
   int ret = 0;

   if( __SIO_OD( fd ).nextb < __SIO_OD( fd ).buf_end ) {
      if( __SIO_OD(fd).buftype == SIO_FULLBUF ) {
         ret = *(__SIO_OD(fd).nextb)++ = (unsigned char) (c);
      } else if ( __SIO_OD(fd).buftype == SIO_LINEBUF ) {
         if( (*(__SIO_OD(fd).nextb) = (unsigned char)(c)) != '\n' ) {
            ret = *(__SIO_OD(fd).nextb)++;
         } else {
            ret = Sputc( fd, *(__SIO_OD(fd).nextb) );
         }
      } else {
         ret = Sputc( fd, c );
      }
   } else {
      ret = Sputc( fd, c );
   }

   return ret;
}


